### PR TITLE
#2617 Support generic functions.

### LINKED ISSUE #2617: [PSyData] Kernel Extraction: Function with generic interfaces cause crash
The following LFRic code causes the driver creation to crash:
```
  USE coordinate_jacobian_mod, ONLY: pointwise_coordinate_jacobian, pointwise_coordinate_jacobian_inverse
...
          jac_inv = pointwise_coordinate_jacobian_inverse(jac, dj)
...
{
      File "/home/joerg/fab-workspace/skeleton-mpif90/source/optimisation/nci-gadi/global.py", line 38, in trans
    extract.apply(kern,{"create_driver": True} )
      File "/home/joerg/work/psyclone/src/psyclone/domain/lfric/transformations/lfric_extract_trans.py", line 152, in apply
    ctu.get_in_out_parameters(nodes, collect_non_local_symbols=True)
      File "/home/joerg/work/psyclone/src/psyclone/psyir/tools/call_tree_utils.py", line 272, in get_in_out_parameters
    self.get_non_local_read_write_info(node_list, read_write_info)
      File "/home/joerg/work/psyclone/src/psyclone/psyir/tools/call_tree_utils.py", line 335, in get_non_local_read_write_info
    return self._resolve_calls_and_unknowns(todo, read_write_info)
      File "/home/joerg/work/psyclone/src/psyclone/psyir/tools/call_tree_utils.py", line 455, in _resolve_calls_and_unknowns
    if sym.is_constant:
    AttributeError: 'GenericInterfaceSymbol' object has no attribute 'is_constant'
}
please check your script
```
Reason is that at line 455 a generic interface is not detected. A generic interface in case of a call is handled way earlier at the caller-side of things (since in case of a call we have a RoutineSymbol), but in case of a function we can't distinguish between an array access and a function call (without looking into the used module).

### PR DESCRIPTION
**Pull request by hiker**
Fixes #2617.

### CODE REVIEW

File: src/psyclone/tests/psyir/tools/call_tree_utils_test.py [hunk 1]
```diff
@@ -128,6 +128,33 @@ def test_call_tree_compute_all_non_locals_non_kernel():
     assert info == []
 
 
+# -----------------------------------------------------------------------------
+@pytest.mark.usefixtures("clear_module_manager_instance")
+def test_call_tree_generic_functions():
+    '''Test handling of generic functions. Each function must be
+    visited (until PSyclone properly resolves generic interfaces)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/tools/call_tree_utils_test.py:135]: As discussed in Teams, please could you clarify slightly. Even when we can do better than we do now, there will still be cases where we can't resolve the target due to a lack of type information.

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks Joerg. Almost perfect. Just one comment needs a bit of clarification.
I'll fire off the integration tests in the meantime.

**Comment by hiker** [comment]
I hope that's better. The CI passed, and given that I only modified a comment, I didn't bother running it again.

**Review by arporter — APPROVED** [review]
New comment is good thanks.
Just waiting on CI now and then I'll merge it.

**Pull request merged**